File: python/spectral.py
import numpy as np
import networkx as nx
import random
import warnings
import pyximport
from scipy.linalg import eigh
from sklearn.cluster import KMeans
pyximport.install(setup_args={
                              "include_dirs":np.get_include()},
                  reload_support=True)

from spectral_aux import Mnn, create_adjacency_matrix


def Laplacian_eigen(G, k):
    n = G.shape[0]
    # Finding Laplacian of a graph
    D = np.zeros((n, n), dtype=np.int)
    for i in range(n):
        D[i, i] = np.sum(G[i, :])
    L = D - G
    _, E = eigh(L, eigvals=(2, k))
    return E


def Mnn_graph(S):
    """
    Graph G is represented by adjacency matrix.
    In graph G vertices i and j are connected (G[i, j] == 1) if and only if
    i is in M nearest neighbours of j or j is in M nearest neighbours of i
    """
    G = create_adjacency_matrix(S)

    # Connecting a graph (randomly)
    G_nx = nx.from_numpy_matrix(G)
    components = list(nx.connected_components(G_nx))
    for i in range(len(components) - 1):
        vertex_1 = random.sample(components[i], 1)[0]
        vertex_2 = random.sample(components[i + 1], 1)[0]
        G[vertex_1, vertex_2] = 1
        G[vertex_2, vertex_1] = 1
    
    return G


def spectral_clustering(X, k, M, verbose=False):
    n = X.shape[0]

    if M > n-1:
        print("spectral clustering: M={} was to large and it was changed to {}".format(M, n-1))
        M = n-1

    if verbose:
        print("Mnn function")
    S = Mnn(X, M)
    
    if verbose:
        print("Mnn_graph function")
    G = Mnn_graph(S)

    if verbose:
        print("Laplacian_eigen function")
    E = Laplacian_eigen(G, k)

    if verbose:
        print("Kmeans")
    kmeans = KMeans(n_clusters=k)
    clustering = kmeans.fit_predict(E)
    
    return clustering

File: R/spectral.R
library(Rcpp)
library(igraph)
sourceCpp(file="spectral_aux.cpp")


Mnn_graph <- function(S){
  "
    Graph G is represented by adjacency matrix.
    In graph G vertices i and j are connected (G[i, j] == 1) if and only if
    i is in M nearest neighbours of j or j is in M nearest neighbours of i
  "
  G <- get_adjacency_matrix(S)
  
  # Connecting a graph
  g_igraph <- graph_from_adjacency_matrix(G)
  components <- clusters(g_igraph)
  for(i in 2:components$no){
    vertex = which.max(components$membership == i)
    G[1, vertex] = 1
    G[vertex, 1] = 1
  }
  return(G)
}


Laplacian_eigen <- function(G, k){
  "Returns eigenenvector matrix of Laplacian of matrix G"
  diagonal <- apply(G, 1, sum)
  D <- diag(diagonal)
  L <- D - G
  n <- nrow(L)
  E <- eigen(L)$vectors[, (n-k):(n-1)]  # eigenvectors corresponding to 2, 3, ... k+1 smallest eigenvalues
  return(E)
}


spectral_clustering <- function(X, k, M){
  "X: matrix of observations we want to cluster
   k: number of clusters
   M: hyperparameter - number of nearest neighbours
  "
  
  S <- Mnn(X, M)
  G <- Mnn_graph(S)
  E <- Laplacian_eigen(G, k)
  km <- kmeans(E, k)
  return(km$cluster)
}

